### ISSUE TITLE
Makefile lacks 'install' target

### ISSUE THREAD

**Issue by alexeymuranov** [gap-system/gap#197]
Currently `Makefile` has no `install` target. Quoting from an email exchange on GAP Support mailing list,
> This means that packagers have to reimplement the equivalent of 'make install' themselves.

I think that for packaging GAP for Unix-like platforms, it would be desirable to have `make install` install everything into appropriate places under `prefix`, preferably in accordance with [Linux Filesystem Hierarchy Standard](http://refspecs.linuxfoundation.org/fhs.shtml).

**Comment by fingolfin** [comment]
Yes, `make install` is not supported. I do agree that it would be desirable in princple, but some work is needed. For starters, GAP itself would need to be made aware of this mode of operation. I spent some time thinking about that in the past. If somebody wants to have a go at this, feel free to ping me to discuss some of the potential pitfalls.

That said, currently GAP is meant to be run directly from its build directory, and it works fine that way, so for me this is not a super high priority issue -- though of course people who want to package GAP for UNIX systems certainly would appreciate it and benefit from it.

**Comment by stevelinton** [comment]
A make install that moves the build directory as a whole to a specified target location and fixes things up to work there (if necessary) would be useful.  However we have no plans to support GAP operating other than from a (or several) GAP root directories so compliance with the filesystem hierarchy standard is not on the cards.

**Comment by frankluebeck** [comment]
I have commented on this topic in [this discussion](https://github.com/gap-system/gap/pull/189#issuecomment-117083595) on a pull request.

Just moving the whole build tree to some other location is not that difficult (but a bit tricky when it is supposed to work including installed packages).

But making GAP work with files distributed, say, according to the Linux Filestystem Hierachy Standard (are there similar standards for non-Linux systems?) would be a lot of work, because all code in GAP and in many packages which needs to read files (say, for loading library and packages, using the help system, reading data files, ...) would need to be checked and in most cases to be changed. It may be difficult to find anyone who is willing to do this work.

**Comment by alexeymuranov** [comment]
About packages, IMO package installation should be separated from the installation of the core system. There is some discussion on the mailing list under subject ["Package Uniformity"](https://mail.gap-system.org/pipermail/gap/2015-June/000070.html) that might be relevant.

**Comment by yurivict** [comment]
There is an ```install``` target that installs some files but not all of them.

This is a weakness of GAP that it doesn't have a correct install target because then packagers have to write their own and this is a more error-prone situation.

FreeBSD has an amendment for the existing install target in post-install [here](https://cgit.freebsd.org/ports/tree/math/gap/Makefile#n77) that seems to be correct. Maybe you can just copy the commands from there.

**Comment by fingolfin** [comment]
> There is an `install` target that installs some files but not all of them.

What are you talking about? The install target does this:
```
install:
        @echo "Error, 'make install' has not yet been implemented"
        exit 1
```

Or are you talking about the various *undocumented* `install-FOO` targets which all start by printing a warning that they are incomplete?

> This is a weakness of GAP that it doesn't have a correct install target because then packagers have to write their own and this is a more error-prone situation.

Agreed.

> 
> FreeBSD has an amendment for the existing install target in post-install [here](https://cgit.freebsd.org/ports/tree/math/gap/Makefile#n77) that seems to be correct. Maybe you can just copy the commands from there.

I can't test it, but it looks neither complete nor fully correct; and it doesn't seems tackle the difficult parts at all. E.g.
- it simply installs `config.h` without care for the ramifications (which might be OK for FreeBSD, if it always installs "from source", but not for distros that ship precompiled binaries) -- I've dealt with that on our master branch, though, so this issue should be resolved / resolvable for 4.12
- it doesn't look like it'll produce a fully working `gac`
- it doesn't adjust compiler flags in sysinfo.gap
- it doesn't deal with packages in a way that would be useful for other distros (though arguably that could be considered as out-of-scope; still, packages need to deal with that)

**Comment by yurivict** [comment]
> What are you talking about? The install target does this:

Sorry, I correct my previous statement: these 2 targets ```install-headers install-libgap``` are used in the FreeBSD port as a base, and then the port-provided ```post-install``` completes it.

The resulting installation passes testsuite.

>I see that you (?) are updating the FreeBSD port, which is great, thanks for working on it.

That's me. -)

**Comment by yurivict** [comment]
Why does ```gac``` need all build directory? Shouldn't installed ```libgap.so``` be sufficient?

**Comment by fingolfin** [comment]
For static linking. But see also https://github.com/gap-system/gap/pull/4778

**Comment by yurivict** [comment]
I see, thanks.

**Comment by fingolfin** [comment]
Of course it was properly implemented, it worked perfectly fine for decades across dozens of platforms (outliving many of those). And long before any desire to support "make install". Sure one could have created an auxiliary libgap.a but that would just shift the issue. Plus these days it would be super confusing because it would have different content than libgap.so (which contains compstat.o, which gac mus exclude)

**Comment by yurivict** [comment]
compstat.o should have been installed in addition to libgap.a, so only these 2 files would have been needed for static linking.

**Comment by yurivict** [comment]
IMO, you should just build libgap.a and keep both static and dynamic linking for gac w/out the need to ever have a build directory.
